[dbquery] Initial contribution (#8780)
openhab/openhab-addons: bundles/org.openhab.binding.dbquery/README.md

# DBQuery Binding

This binding allows creating items from the result of native database queries.

> It currently only supports InfluxDB 2.0.

Reviewer: 2.0 or 2.x ?

> You can use the addon in any situation where you want to create an item from a native query, the source of the query can
> be any supported database, and don't need to be the one you use as persistence service in openHAB.

Reviewer: ```suggestion
You can use the addon in any situation where you want to create an item from a native query.
The source of the query can be any supported database, and doesn't need to be the one you use as the persistence service in openHAB.
```

Some use cases can be:

- Integrate a device that stores its data in a database
- Query derived data from you openHAB persistence, for example with Influx2 tasks you can process your data to create a new one   
- Bypass limitations of current openHAB persistence queries

> 

Reviewer: ```suggestion
```
I think the code analysis report will flag this. Only one empty line.

## Supported Things

There are two types of supported things: `influxdb2` and a `query`.

> For each different database, you want to connect you must define a `Bridge` thing for that database.

Reviewer: ```suggestion
For each different database you want to connect to, you must define a `Bridge` thing for that database.
```

> Then each `Bridge` can define as much as wanted `Query` things you want to execute.

Reviewer: ```suggestion
Then each `Bridge` can define as many `Query` things that you want to execute.
```

> Current supported `Bridge` is:
> 
> - `influxdb2`
> 

Reviewer: ```suggestion
```
Not needed as the text above and below are enough.

| bucket       | Yes      | database bucket name                      |

### query

The `Query` thing defines a native query that provides several channels that you can bind to items. 

#### Query parameters

The query items support the following parameters:

| Parameter    | Required | Default  | Description                                                           |
|--------------|----------|----------|-----------------------------------------------------------------------|
| query        | true     |          | Query string in native syntax                                         |
| interval     | false    | 0        | Interval in seconds in which the query is automatically executed      |
| hasParameters| false    | false    | True if the query has parameters, false otherwise                         | 
| timeout      | false    | 0        | Query execution timeout in seconds                                    |
| scalarResult | false    | true     | If query always returns a single value or not                         |
| scalarColumn | false    |          | In case of multiple columns, it indicates which to use for scalarResult|

> that are described further in the following subsections

Reviewer: ```suggestion
These are described further in the following subsections
```

##### query  

The query the items represents in the native language of your database:

 - Flux for `influxdb2`
 
#### hasParameters

If `hasParameters=true` you can use parameters in the query string that can be dynamically set with the `setQueryParameters` action.
 
 For InfluxDB use the `${paramName}` syntax for each parameter, and keep in mind that the values from that parameters must be from a trusted source as current
 parameter substitution is subject to query injection attacks.
 
#### timeout

> A time-out in seconds to wait for the query result, if it's exceeded, result will be discarded and the addon will do its best effort to cancel the query.

Reviewer: ```suggestion
A time-out in seconds to wait for the query result, if it's exceeded, the result will be discarded and the addon will do its best to cancel the query.
```

#### scalarResult 

If `true` the query is expected to return a single scalar value that will be available to `result` channels as string, number, boolean,...
If the query can return several rows and/or several columns per row then it needs to be set to `false` and the result can be retrieved in `resultString`
channel as JSON or using the `getLastQueryResult` action.   

#### scalarColumn

In case `scalarResult` is `true` and the select returns multiple columns you can use that parameter to choose which column to use to extract the result.

## Channels

Query items offer the following channels to be able to query / bind them to items:

| Channel Type ID | Item Type | Description                                                                                                                        |
|-----------------|-----------|------------------------------------------------------------------------------------------------------------------------------------|
| execute         | Switch    | Send `ON` to execute the query manually. It also indicates if query is currently running (`ON`) or not running (`OFF`)          |

> | resultString    | String    | Result of last executed query as an String |

Reviewer: ```suggestion
| resultString    | String    | Result of last executed query as a String |
```

| resultNumber    | Number    | Result of last executed query as a Number, query must have `scalarResult=true` |
| resultDateTime  | DateTime  | Result of last executed query as a DateTime, query must have `scalarResult=true` |
| resultContact   | Contact   | Result of last executed query as Contact, query must have `scalarResult=true` |
| resultSwitch    | Switch    | Result of last executed query as Switch, query must have `scalarResult=true` |
| parameters      | String    | Contains parameters of last executed query as JSON|

> | correct         | Switch    | `ON` if the result of last executed query was correct |

Reviewer: Can you make this clearer as I am not sure how this is different to resultSwitch channel? Do you mean `correct`, or do you mean true like a boolean?

Reviewer: So do I understand it if I suggest this wording?
```
`ON` if the last executed query completed successfully, `OFF` if the query failed.
```
I feel the description does not explain what the channel is for when someone is new to the binding, perhaps I am a slow learner :)

Author: Thanks, sounds better

All the channels, except `execute`, are updated when the query execution finishes, and while there is a query in execution they have the values from
last previous executed query.

The `resultString` channel is the only valid one if `scalarResult=false`, and in that case it contains the query result serialized to JSON in that format:

    {
        correct : true,
        data : [
            { 
                column1 : value,
                column2 : value
            },
            { ... }, //row2        
            { ... }  //row3
        ]
    }
    
### Channel Triggers

#### calculateParameters

Triggers when there's a need to calculate parameters before query execution.

> When a query has `hasParameters=true` it fires the `calculateParameters` channel trigger and pause the execution until `setQueryParameters` action is call in

Reviewer: ```suggestion
When a query has `hasParameters=true` it fires the `calculateParameters` channel trigger and pauses the execution until `setQueryParameters` action is call in
```

 that query.
 

> In the case a query has parameters it's expected that there is some rule that catch the `calculateParameters` trigger, calculate the parameters
>  with the corresponding logic and then call the `setQueryParameters` action, after that the query will be really executed.

Reviewer: ```suggestion
In the case a query has parameters, it's expected that there is a rule that catches the `calculateParameters` trigger, calculate the parameters
 with the corresponding logic and then calls the `setQueryParameters` action, after that the query will be executed.
```

- String query: The query to execute
- Map<String,Object>: Query parameters (empty map if not needed)
- int timeout: Query timeout in seconds

And it returns an `ActionQueryResult` that has the following properties:

- correct (boolean) : True if the query was executed correctly, false otherwise
- data (List<Map<String,Object>>): A list where each element is a row that is stored in a map with (columnName,value) entries  
- isScalarResult: It returns if the result is scalar one (only one row with one column)
- resultAsScalar: It returns the result as a scalar if possible, if not returns null

Example (using Jython script):

     from core.log import logging, LOG_PREFIX 
     log = logging.getLogger("{}.action_example".format(LOG_PREFIX))
     map = {"time" : "-2h"}
     influxdb = actions.get("dbquery","dbquery:influxdb2:sampleQuery") //Get bridge thing
     result = influxdb.executeQuery("from(bucket: \"default\") |> range(start:-2h)  |> filter(fn: (r) => r[\"_measurement\"] == \"go_memstats_frees_total\")  |> filter(fn: (r) => r[\"_field\"] == \"counter\")  |> mean()",{},5)
     log.info("execute query result is "+str(result.data))
    

> Use this action with care, because as the query is executed synchronously is not good to execute long-running queries and block script execution.

Reviewer: ```suggestion
Use this action with care, because as the query is executed synchronously, it is not good to execute long-running queries that can block script execution.
```

### For Queries

#### setQueryParameters

> It's used for queries with parameters to set them. To execute the action you need to pass the parameters as a Map.

Reviewer: ```suggestion
It's used for queries with parameters to set them.
To execute the action you need to pass the parameters as a Map.
```

Example (using Jython script):

    params = {"time" : "-2h"}
    dbquery = actions.get("dbquery","dbquery:query:queryWithParams")  //Get query thing
    dbquery.setQueryParameters(params)

#### getLastQueryResult

> It can be used in scripts to get the last query result. It doesn't have any parameters and returns an `ActionQueryResult` as defined in `executeQuery` action.

Reviewer: ```suggestion
It can be used in scripts to get the last query result.
It doesn't have any parameters and returns an `ActionQueryResult` as defined in `executeQuery` action.
```

Example (using Jython script):

    dbquery = actions.get("dbquery","dbquery:query:queryWithParams")  //Get query thing
    result = dbquery.getLastQueryResult()
    

## Examples

### The Simplest case 

> Define a InfluxDB2 database thing and a query with an interval execution. That executes the query every 15 seconds and punts the result in `myItem`

Reviewer: ```suggestion
Define a InfluxDB2 database thing and a query with an interval execution.
That executes the query every 15 seconds and punts the result in `myItem`.
```
New line for each new sentence

# DBQuery Binding

This binding allows creating items from the result of native database queries.

> It currently supports InfluxDB 2.0 and will support major relational databases through JDBC.

Reviewer 2: If you want to release it without JDBC support first, better don't mention it here.

Author: Ok, I will remove it. 
If I found some time, and now that you are reviewing it, and I hope it can be accepted after making desired corrections. I hope to work on the SQL binding soon but I think it's better to be able to integrate current work just In case I found little time and it comes later.

In fact, I want to try TimescaleDB for measurements and will need to implement that part because being able to integrate iotawatt readings is a must for me.

You can use the addon in any situation where you want to create an item from a native SQL query, the source of the query can
be any supported database, and don't need to be the one you use as persistence service in InfluxDB.
Some use cases can be:

> - Integrate a device that stores it's data in a database

Reviewer 2: ```suggestion
- Integrate a device that stores its data in a database
```

> - Query derived data from you OpenHAB persistence, for example with Influx2 tasks you can process your data to create a new one   

Reviewer 2: ```suggestion
- Query derived data from you openHAB persistence, for example with Influx2 tasks you can process your data to create a new one   
```

- Bypass limitations of current OpenHAb persistence queries

## Supported Things

> There are two types of supported things: `DatabaseBridge` and a `Query`.

Reviewer 2: Thing type UIDs should start with lower case. Also, the real Thing Type UIDs should be mentioned here.
```suggestion
There are two types of supported things: `Influx2Bridge` and `Query`.
```

| hasParameters| false    | false    | True if the query has parameters, false otherwise                         | 
| timeout      | false    | 0        | Query execution timeout in seconds                                    |
| scalarResult | false    | true     | If query always returns a single value or not                         |
| scalarColumn | false    |          | In case of multiple columns, it indicates which to use for scalarResult|

that are described further in the following subsections

##### query  

The query the items represents in the native language of your database:

 - Flux for `Influx2Bridge`
 - SQL for `JDBCBridge`
 
#### hasParameters

If `hasParameters=true` you can use parameters in the query string that can be dynamically set with the `setQueryParameters` action.
 
 For InfluxDB use the `${paramName}` syntax for each parameter, and keep in mind that the values from that parameters must be from a trusted source as current
 parameter substitution is subject to query injection attacks.
 
#### timeout

A time-out in seconds to wait for the query result, if it's exceeded, result will be discarded and the addon will do it's best effort to cancel the query.

> Currently it's ignored and it will be implemented in a future version.

Reviewer 2: If it's not implemented, better remove it from here.

Query items offer the following channels to be able to query / bind them to items:

| Channel Type ID | Item Type | Description                                                                                                                        |
|-----------------|-----------|------------------------------------------------------------------------------------------------------------------------------------|
| execute         | Switch    | Send `ON` to execute the query manually. It also indicates if query is currently running (`ON`) or not running (`OFF`)          |
| resultString    | String    | Result of last executed query as an String |
| resultNumber    | Number    | Result of last executed query as a Number, query must have `scalarResult=true` |
| resultDateTime  | DateTime  | Result of last executed query as a DateTime, query must have `scalarResult=true` |
| resultContact   | Contact   | Result of last executed query as Contact, query must have `scalarResult=true` |
| resultSwitch    | Switch    | Result of last executed query as Switch, query must have `scalarResult=true` |
| parameters      | String    | Contains parameters of last executed query as JSON|

> | correct         | Switch    | True if the result of last executed query was correct |

Reviewer 2: ```suggestion
| correct         | Switch    | ON if the result of last executed query was correct |
```
